crates/allocator/src/bump.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple bump allocator.
+//!
+//! It's goal to have a much smaller footprint than the admittedly more full-featured `wee_alloc`
+//! allocator which is currently being used by ink! smart contracts.
+//!
+//! The heap which will be used by this allocator is a single page of memory, which in Wasm is
+//! 64KiB. We do not expect contracts to use more memory than this (for now), so we will throw an
+//! OOM error instead of requesting more memory.
+
+use core::alloc::{
+    GlobalAlloc,
+    Layout,
+};
+
+/// A page in Wasm is 64KiB
+const PAGE_SIZE: usize = 64 * 1024;
+
+static mut INNER: InnerAlloc = InnerAlloc::new();
+
+pub struct BumpAllocator;
+
+unsafe impl GlobalAlloc for BumpAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        INNER.alloc(layout)
+    }
+
+    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+}
+
+struct InnerAlloc {
+    /// Points to the start of the next available allocation.
+    next: usize,
+    /// We need some way to initialize our heap. However, I can't figure out how to get the
+    /// initialization working properly with `lazy_static` so this hack is the best I got for now.
+    heap_initialized: bool,
+}
+
+impl InnerAlloc {
+    pub const fn new() -> Self {
+        Self {
+            next: 0,
```

Comment:
This is wrong. You will most probably overwrite statically allocated data structures put there by the linker (data, stack). You can't assume that your heap starts at `0`. When I compile the flipper contract I can see that the initial memory size is `2` pages and those are in use:

```wat
(import "env" "memory" (memory (;0;) 2 16))
````


You have two choices now:

1. Use the return value of `memory_grow` to calculate where your heap should start. This is a freshly allocated page and guaranteed to not be in used by the linker. This is how `wee_alloc` deals with it.

2. Try to be clever and get away without even calling `memory.grow` once. AFAIK `lld` used to expose the `__heap_base` symbol which you can reference to determine the address where you can safely put your heap (at this address and any higher address). However, it stopped to to do so by default (you need to put `-Clink-arg=--export=__heap_base` into your rust flags now). Also, one need to make sure to dimension the initial memory large enough so that there is enough space for a heap. In any case this needs new flags to be supplied to the linker (therefore changes to cargo contract). I would recommend using the first approach first and leave this as a later optimization. By the way this is how substrate runtimes handle their heap. With the little difference that the allocator is placed inside the client and not the runtime.

Reply:
Oh wow yeah, good point! I think I took the whole "each contract has its own Wasm sandbox" thing too seriously and forgot about everything else 😅

> When I compile the flipper contract I can see that the initial memory size is 2 pages and those are in use

Where/how did you get this? I looked through the "normal" build artifacts (`*.contract`, `*.wasm`, `metadata.json`) and can't find it

> Use the return value of memory_grow to calculate where your heap should start. This is a freshly allocated page and guaranteed to not be in use by the linker. This is how wee_alloc deals with it.

Yeah, this makes sense.

> Try to be clever and get away without even calling memory.grow once.

I'm gonna skip being clever and get something working first 😄

Reply:
`Where/how did you get this? I looked through the "normal" build artifacts (*.contract, *.wasm, metadata.json) and can't find it`

Just disassemble the `.wasm` file. You can use `wasm2wat` for that. Is it part of the `wabt` package.